x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/LinearProgrammingPlanSolver.java

```diff
@@ -0,0 +1,341 @@
+ */
+class LinearProgrammingPlanSolver {
+
+    private static final Logger logger = LogManager.getLogger(LinearProgrammingPlanSolver.class);
+
+    private static final long RANDOMIZATION_SEED = 738921734L;
+    private static final double L1 = 0.9; // Must be a number < 1. 0.9 gave good results during testing.
+    private static final double INITIAL_W = 0.2;
+    private static final int RANDOMIZED_ROUNDING_ROUNDS = 20; // more rounds do not seem to improve things.
+
+    /**
+     * Ojalgo creates a 2D double array with size of complexity {@link #memoryComplexity()}.
+     * We set a limit of storing 10M doubles (~80MB) to avoid OOM. We fall back to the
+     * bin packing solution if we're over that limit.
+     */
+    private static final int MEMORY_COMPLEXITY_LIMIT = 10_000_000;
+
+    private final Random random = new Random(RANDOMIZATION_SEED);
+
+    private final List<Node> nodes;
+    private final List<Model> models;
+    private final Map<Node, Double> normalizedMemoryPerNode;
+    private final Map<Node, Integer> coresPerNode;
+    private final Map<Model, Double> normalizedMemoryPerModel;
+
+    private final int maxNodeCores;
+    private final long maxModelMemoryBytes;
+
+    LinearProgrammingPlanSolver(List<Node> nodes, List<Model> models) {
+        this.nodes = nodes;
+
+        // Filter out models that are not already assigned and do not fit on any node
+        long maxNodeMemory = nodes.stream().map(Node::availableMemoryBytes).max(Long::compareTo).orElse(0L);
+        this.models = models.stream()
+            .filter(m -> m.currentAllocationByNodeId().isEmpty() == false || m.memoryBytes() <= maxNodeMemory)
+            .toList();
+
+        maxNodeCores = this.nodes.stream().map(Node::cores).max(Integer::compareTo).orElse(0);
+        maxModelMemoryBytes = this.models.stream().map(Model::memoryBytes).max(Long::compareTo).orElse(0L);
+        normalizedMemoryPerNode = this.nodes.stream()
+            .collect(Collectors.toMap(Function.identity(), n -> n.availableMemoryBytes() / (double) maxModelMemoryBytes));
+        coresPerNode = this.nodes.stream().collect(Collectors.toMap(Function.identity(), Node::cores));
+        normalizedMemoryPerModel = this.models.stream()
+            .collect(Collectors.toMap(Function.identity(), m -> m.memoryBytes() / (double) maxModelMemoryBytes));
+    }
+
+    AssignmentPlan solvePlan() {
+        if (models.isEmpty() || maxNodeCores == 0) {
+            return AssignmentPlan.builder(nodes, models).build();
+        }
+
+        Tuple<Map<Tuple<Model, Node>, Double>, AssignmentPlan> weightsAndBinPackingPlan = calculateWeightsAndBinPackingPlan();
+
+        Map<Tuple<Model, Node>, Double> allocationValues = new HashMap<>();
+        Map<Tuple<Model, Node>, Double> assignmentValues = new HashMap<>();
+        if (solveLinearProgram(weightsAndBinPackingPlan.v1(), allocationValues, assignmentValues) == false) {
+            return weightsAndBinPackingPlan.v2();
+        }
```

Comment:
Is it possible to know that you cannot get any better than the `weightsAndBinPackingPlan`? It seems like a big waste to run the linear program and THEN pick the bin packing solution.

Reply:
I don't think there's a way to know the bin-packing solution is better before running the solver. At least not without adding complexity and I'm not sure it's worth it given the runtime of this and the fact that it only happens on specific events (model starts, nodes joining, etc.). Also, `solveLinearProgram` may return `false` which means the solver couldn't find a feasible solution.

Reply:
>  is better

I am thinking that we determine that it is "good enough" like we have some acceptable amount of waste.

But sure, I understand it's a premature optimization. I am just slightly concerned over the fact that this is running on a master node with a bunch of other things. 

I suppose we just need to be careful that plans are not spammed in a "rolling upgrade" scenario or something (which is a separate optimization